docs/bigquery/snippets.py

```diff
@@ -2579,6 +2579,178 @@ def test_query_no_cache(client):
+    job_config.table_definitions = {table_id: external_config}
+
+    # Example query to find states starting with 'W'
+    sql = 'SELECT * FROM {} WHERE name like "W%"'.format(table_id)
+
+    query_job = client.query(sql, job_config=job_config)  # API request
+
+    w_states = list(query_job)  # Waits for query to finish
+    print('There are {} states with names starting with W.'.format(
+        len(w_states)))
+    # [END bigquery_query_external_gcs_temp]
+    assert len(w_states) == 4
+
+
+def test_query_external_gcs_permanent_table(client, to_delete):
+    dataset_id = 'query_external_gcs_{}'.format(_millis())
+    dataset = bigquery.Dataset(client.dataset(dataset_id))
+    client.create_dataset(dataset)
+    to_delete.append(dataset)
+
+    # [START bigquery_query_external_gcs_perm]
+    # from google.cloud import bigquery
+    # client = bigquery.Client()
+    # dataset_id = 'my_dataset'
+
+    # Configure the external data source
+    dataset_ref = client.dataset(dataset_id)
+    table_id = 'us_states'
+    schema = [
+        bigquery.SchemaField('name', 'STRING'),
+        bigquery.SchemaField('post_abbr', 'STRING')
+    ]
+    table = bigquery.Table(dataset_ref.table(table_id), schema=schema)
+    external_config = bigquery.ExternalConfig('CSV')
+    external_config.source_uris = [
+        'gs://cloud-samples-data/bigquery/us-states/us-states.csv',
+    ]
+    external_config.options.skip_leading_rows = 1  # optionally skip header row
+    table.external_data_configuration = external_config
+
+    # Create a permanent table linked to the GCS file
+    table = client.create_table(table)  # API request
+
+    # Example query to find states starting with 'W'
+    sql = 'SELECT * FROM {}.{} WHERE name like "W%"'.format(
+        dataset_id, table_id)
+
+    query_job = client.query(sql)  # API request
+
+    w_states = list(query_job)  # Waits for query to finish
+    print('There are {} states with names starting with W.'.format(
+        len(w_states)))
+    # [END bigquery_query_external_gcs_perm]
+    assert len(w_states) == 4
+
+
+def test_query_external_sheets_temporary_table(client):
+    # [START bigquery_query_external_sheets_temp]
+    # [START bigquery_auth_drive_scope]
+    # from google.cloud import bigquery
+
+    import google.auth
+
+    # Create credentials with Drive & BigQuery API scopes
+    # Both APIs must be enabled for your project before running this code
+    credentials, project = google.auth.default(scopes=[
+        'https://www.googleapis.com/auth/drive',
+        'https://www.googleapis.com/auth/bigquery',
+    ])
+    client = bigquery.Client(credentials=credentials)
+    # [END bigquery_auth_drive_scope]
+
+    # Configure the external data source and query job
+    external_config = bigquery.ExternalConfig('GOOGLE_SHEETS')
+    # Use a shareable link or grant viewing access to the email address you
+    # used to authenticate with BigQuery (this example Sheet is public)
```

Comment:
👍 Nice sheet.

Should we clarify that it is possible to share sheets with service accounts?

Reply:
I didn't want to confuse non-service account users by referencing service accounts since those would also be email addresses. Do you think it is necessary for clarity to say that this also applies to service accounts?

Reply:
When I search for "google sheets share with service account" I don't see any official docs that verify that you can do it, but you're right that email is the most general.

Probably the most appropriate place for how to share with a service account is in the Sheets API or Drive API docs rather than this sample.

Reply:
When I was testing out how this worked, I just shared the file with my service account email address and it worked. For example, share with `my-service-account@project.iam.gserviceaccount.com` instead of `actual.person@gmail.com`. The doc that this sample will be included in actually recommends using a shareable link, which I also referenced in the comment, and is probably a better idea anyway. I just thought it would be nice to document the options.

Reply:
Shareable link allows service accounts to access for docs owned by gmail accounts but not for docs owned by gsuite customers.